### ISSUE TITLE
[qt5] build failure on Windows

### ISSUE THREAD

**Issue by dannypike** [microsoft/vcpkg#21343]
**Host Environment**
- OS: Windows 10
- Compiler: VS 2019

**To Reproduce**
`./vcpkg install qt5`

**Failure logs** 
```
[...]/qt5-svg/src/5.15.2-57a185b127.clean/.qmake.conf:1: Cannot find feature qt_build_config
Could not find qmake spec 'win32-msvc'.
Error processing project file: U:\\qt5-svg\src\5.15.2-57a185b127.clean\qtsvg.pro
```

I upgraded this morning and the build now fails. It was working a few days ago.

**Comment by JackBoosY** [comment]
Please remove `qt5-base` and try again.

Thanks.

**Comment by dannypike** [comment]
It now fails trying to build `qt-imageformats`.

```
D:\work\github\vcpkg>set VCPKG
VCPKG_DEFAULT_TRIPLET=x64-windows
VCPKG_KEEP_ENV_VARS=VSCMD_SKIP_SENDTELEMETRY
VCPKG_MAX_CONCURRENCY=6
VCPKG_ROOT=D:\work\github\vcpkg

D:\work\github\vcpkg>vcpkg remove qt5
The following packages are not installed, so not removed:
    qt5:x64-windows
Package qt5:x64-windows is not installed

D:\work\github\vcpkg>git pull
Already up to date.

D:\work\github\vcpkg>vcpkg install qt5
Computing installation plan...
The following packages will be built and installed:
    qt5[activeqt,core,declarative,essentials,imageformats,multimedia,networkauth,quickcontrols2,svg,tools]:x64-windows -> 5.15.2#1
  * qt5-activeqt[core]:x64-windows -> 5.15.2#1
  * qt5-declarative[core]:x64-windows -> 5.15.2#2
  * qt5-imageformats[core]:x64-windows -> 5.15.2#1
  * qt5-multimedia[core]:x64-windows -> 5.15.2#1
  * qt5-networkauth[core]:x64-windows -> 5.15.2#1
  * qt5-quickcontrols2[core]:x64-windows -> 5.15.2#1
  * qt5-svg[core]:x64-windows -> 5.15.2#2
  * qt5-tools[core]:x64-windows -> 5.15.2#1
Additional packages (*) will be modified to complete this operation.
Detecting compiler hash for triplet x64-windows...
Restored 0 packages from D:\Users\peter\AppData\Local\vcpkg\archives in 1.192 ms. Use --debug to see more details.
Starting package 1/9: qt5-imageformats:x64-windows
Building package qt5-imageformats[core]:x64-windows...
CMake Warning at scripts/cmake/vcpkg_buildpath_length_warning.cmake:18 (message):
  qt5-imageformats's buildsystem uses very long paths and may fail on your
  system.

  We recommend moving vcpkg to a short path such as 'C:\src\vcpkg' or using
  the subst command.
Call Stack (most recent call first):
  installed/x64-windows/share/qt5/qt_download_submodule.cmake:13 (vcpkg_buildpath_length_warning)
  installed/x64-windows/share/qt5/qt_submodule_installation.cmake:4 (qt_download_submodule)
  ports/qt5-imageformats/portfile.cmake:54 (qt_submodule_installation)
  scripts/ports.cmake:142 (include)


-- Using cached qtimageformats-everywhere-src-5.15.2.tar.xz.
-- Cleaning sources at D:/work/github/vcpkg/buildtrees/qt5-imageformats/src/5.15.2-bdc20f8631.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source D:/work/github/vcpkg/downloads/qtimageformats-everywhere-src-5.15.2.tar.xz
-- Using source at D:/work/github/vcpkg/buildtrees/qt5-imageformats/src/5.15.2-bdc20f8631.clean
-- Using cached msys-mingw-w64-i686-pkg-config-0.29.2-2-any.pkg.tar.zst.
-- Using cached msys-mingw-w64-i686-libwinpthread-git-8.0.0.5906.c9a21571-1-any.pkg.tar.zst.
-- Using msys root at D:/work/github/vcpkg/downloads/tools/msys2/aa5af7b2aa7e90e8
-- Configuring x64-windows-rel
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:127 (message):
    Command failed: D:/work/github/vcpkg/installed/x64-windows/tools/qt5/bin/qmake.exe CONFIG-=debug CONFIG+=release CONFIG-=static CONFIG*=shared D:/work/github/vcpkg/buildtrees/qt5-imageformats/src/5.15.2-bdc20f8631.clean -qtconf D:/work/github/vcpkg/buildtrees/qt5-imageformats/x64-windows-rel/qt.conf -- -system-tiff -system-webp -jasper -no-mng -verbose WEBP_INCDIR=D:/work/github/vcpkg/installed/x64-windows/include "TIFF_LIBS=D:/work/github/vcpkg/installed/x64-windows/lib/tiff.lib D:/work/github/vcpkg/installed/x64-windows/lib/lzma.lib D:/work/github/vcpkg/installed/x64-windows/lib/jpeg.lib D:/work/github/vcpkg/installed/x64-windows/lib/zlib.lib" "WEBP_LIBS=D:/work/github/vcpkg/installed/x64-windows/lib/webpdecoder.lib D:/work/github/vcpkg/installed/x64-w
indows/lib/webpdemux.lib D:/work/github/vcpkg/installed/x64-windows/lib/libwebpmux.lib D:/work/github/vcpkg/installed/x64-windows/lib/webp.lib" "JASPER_LIBS=D:/work/github/vcpkg/installed/x64-windows/lib/jasper.lib D:/work/github/vcpkg/installed/x64-windows/lib/freeglut.lib D:/work/github/vcpkg/installed/x64-windows/lib/jpeg.lib D:/work/github/vcpkg/installed/x64-windows/lib/zlib.lib"
    Working Directory: D:/work/github/vcpkg/buildtrees/qt5-imageformats/x64-windows-rel
    Error code: 3
    See logs for more information:
      D:\work\github\vcpkg\buildtrees\qt5-imageformats\config-x64-windows-rel-err.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_configure_qmake.cmake:87 (vcpkg_execute_required_process)
  installed/x64-windows/share/qt5/qt_build_submodule.cmake:10 (vcpkg_configure_qmake)
  installed/x64-windows/share/qt5/qt_submodule_installation.cmake:8 (qt_build_submodule)
  ports/qt5-imageformats/portfile.cmake:54 (qt_submodule_installation)
  scripts/ports.cmake:142 (include)


Error: Building package qt5-imageformats:x64-windows failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `git pull` and `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  package: qt5-imageformats[core]:x64-windows -> 5.15.2#1
  vcpkg version: 2021-11-02-af04ebf6274fd6f7a941bff4662b3955c64f6f42
  vcpkg-tool version: 897ff9372 2021-11-11 (10 hours ago)

Additionally, attach any relevant sections from the log files above.

D:\work\github\vcpkg>type D:\work\github\vcpkg\buildtrees\qt5-imageformats\config-x64-windows-rel-err.log
D:/work/github/vcpkg/buildtrees/qt5-imageformats/src/5.15.2-bdc20f8631.clean/.qmake.conf:1: Cannot find feature qt_build_config
Could not find qmake spec 'win32-msvc'.
Error processing project file: D:\work\github\vcpkg\buildtrees\qt5-imageformats\src\5.15.2-bdc20f8631.clean\qtimageformats.pro

D:\work\github\vcpkg>
```

**Comment by dannypike** [comment]
Did you update the port, because the `git pull` didn't do anything?

**Comment by JackBoosY** [comment]
> Did you update the port, because the `git pull` didn't do anything?

No, but it maybe out-dated.

Can you please provide _ D:\work\github\vcpkg\buildtrees\qt5-imageformats\config-x64-windows-rel-err.log_ and _ D:\work\github\vcpkg\buildtrees\qt5-imageformats\config-x64-windows-rel-out.log_?

**Comment by dannypike** [comment]
Ah, no. It definitely is up-to-date. That's the reason why it is now failing, when it has been working OK for a while - I have just upgraded `vcpkg` to the latest ports.

Here is the -err.log:
```
D:/work/github/vcpkg/buildtrees/qt5-imageformats/src/5.15.2-bdc20f8631.clean/.qmake.conf:1: Cannot find feature qt_build_config
Could not find qmake spec 'win32-msvc'.
Error processing project file: D:\work\github\vcpkg\buildtrees\qt5-imageformats\src\5.15.2-bdc20f8631.clean\qtimageformats.pro
```
it's the same error as in my original post. 

The -out.log is empty:
```
D:\work\github\vcpkg\buildtrees\qt5-imageformats>dir config-x64-windows-rel-*.log
 Volume in drive D is Coriander
 Volume Serial Number is 9C74-7C18

 Directory of D:\work\github\vcpkg\buildtrees\qt5-imageformats

12/11/2021  08:12               298 config-x64-windows-rel-err.log
12/11/2021  08:12                 0 config-x64-windows-rel-out.log
               2 File(s)            298 bytes
               0 Dir(s)  238,651,023,360 bytes free
```

presumably because it didn't even start the build, because is cannot find the feature, as per the error message.

**Comment by dannypike** [comment]
I noticed that there were more commits to `vcpkg`, so I tried a new rebuild and the error has changed. It's now failing when it tries to build qt5-declarative:

```
Running configuration tests...
Project ERROR: Unknown feature object c++14 in expression 'features.c++14 || tests.cxx14_make_unique'.
```

I've tried deleting the whole `vcpkg` folder (all 120 GBytes of it!) and pulling everything again. It still fails.

Is there something else I need to do to start with a completely fresh `vcpkg` setup?

**Comment by dannypike** [comment]
@JackBoos Apparently, the problem seems to have arisen when I installed Visual Studio 2022. I can build it on a machine that has only ever seen VS2019, but it breaks when you also have VS2022 installed.

Should I create a separate ticket for that? If so, it is a problem with VS2022 or with `vcpkg`, do you think?

**Comment by JackBoosY** [comment]
@dannypike I will look into it.

**Comment by JackBoosY** [comment]
Please also notice that you should use the same compiler to build all ports that include dependencies.

**Comment by dannypike** [comment]
Yes, thank you; I assumed that was the case.

I tried swapping the whole installation between 142 and 143 by using `vcpkg remove` on all the packages, deleting the `buildtrees` folder and deleting the cache at `%APPDATA%\vcpkg`. When that didn't work, I tried creating a brand new VM (using Hyper-V) and installing only VS2019. That worked this morning, so I posted my comment above.

I'm currently trying to make my machine forget that I ever had VS2022 on it, so that I can get going again while you look into this. :)

**Comment by JackBoosY** [comment]
We couldn't reproduce this issue in Visual Studio 2022 release now.

**Comment by dannypike** [comment]
It's quite repeatable for me. Can you confirm how you have tried to reproduce it?

I tried this morning with a new Windows 10 virtual machine. Here's a simplified version of what I did

1. Installed Visual Studio Community 2022 with C++, and MFC support using the installer that I  downloaded from the Microsoft web site (in my case it was called `vs_community__ee9930e5.8353.4c24.85c4.efca820bafe5.exe`).
2. Downloaded `vcpkg` from GitHub, using: `git clone https://github.com/microsoft/vcpkg.git` command.
3. Tried to install one of the Qt5 packages into the new empty vcpkg installation using: `vcpkg install qt5-declarative`

That fails for me with the error that I gave earlier:
```
Project ERROR: Unknown feature object c++14 in expression 'features.c++14 || tests.cxx14_make_unique'.
```

In summary:

1. I can compile Qt5 on a VM that has never seen VS2022;
2. I cannot compile Qt5 on a VM that has seen VS2022, whether it had previously seen VS2019 or not;
3. I cannot compile Qt5 on a VM that has had VS2022 installed and then removed, or that has both VS2019 and VS2022 and had both removed and just VS2019 reinstalled.

How did you test your setup? What part of Qt5 did you try to build?

**Comment by dannypike** [comment]
@JackBoosY While we're trying to sort this out, can you please give me access to reinstall Community 2019?

I'm stuck being unable to compile Qt5, because the installer insists on installing only v2022 and I can't use that because of this problem.

**Comment by JackBoosY** [comment]
@dannypike Of couse, but please note that if you installed both Visual Studio 2019 and Visual Studio 2020, vcpkg will always choose the newer one.

**Comment by dannypike** [comment]
Yes. I am in the process of removing everything associated with VS2022 from my machine (registry etc), to try to get back to a working setup.

Once I have done that, I will install VS2019 and, hopefully, have eliminated the problem.

**Comment by dannypike** [comment]
@JackBoosY Can you please tell me how you tested it?

It fails for me with a brand new VM, with a new copy of VS2022. What test did you do that worked for you?

**Comment by dannypike** [comment]
@JackBoosY What is happening with this, please?

**Comment by LilyWangLL** [comment]
@dannypike I used Visual Studuo Enterprise 2022 d17.0 to install qt5-imageformats:x64-windows succeed. 
![image](https://user-images.githubusercontent.com/94091114/142980270-92102196-4ad2-4e91-a3df-91d3ad20c2a9.png)
![image](https://user-images.githubusercontent.com/94091114/142980343-77cd5e4b-90ad-49e9-8748-5444cafc97f1.png)
```
Installing package qt5-imageformats[core]:x64-windows...
Elapsed time for package qt5-imageformats:x64-windows: 584.4 ms

Total elapsed time: 7.09 s

The package qt5-imageformats provides CMake targets:

    find_package(Qt5Gui CONFIG REQUIRED)
    # Note: 2 target(s) were omitted.
    target_link_libraries(main PRIVATE Qt5::QJp2Plugin Qt5::QTgaPlugin Qt5::QICNSPlugin Qt5::QTiffPlugin)
```

**Comment by JackBoosY** [comment]
Also, can you please provide the full log _config-x64-windows-rel-*.log_?
I can make sure there is a qt log in _D:\work\github\vcpkg\buildtrees\qt5-imageformats\x64-windows-dbg_, can you please provide that?

**Comment by dannypike** [comment]
@LilyWangLL Thank you for your reply.

Yes, `qt5-imageformats` starting working for me after the repo was updated (please see my comment from 10 days ago, 13Nov2021 05:20). The problem is now reproducible by trying to compile `qt5-declarative`, instead. Can you try that, please?

If that works for you, please understand that it builds for me only with `VS2019 v16.11.5`. If I try with `VS2019 v16.11.7`, then it fails with this error. I notice that you are using a version of VS2022 that is out of date and not the same as mine (I am using v17.9.1). Given that it only appeared in VS2019 v16.11.7, I would be interested to know if the problem appears once you upgrade to the latest version of VS2022.

**Comment by dannypike** [comment]
@JackBoosY Thank you for your reply. `qt5-imageformats` has been building OK since I upgraded to the latest port on 13Nov2021 (as per my comment above). The problem can now been seen by trying to build `qt5-declarative`.

I have already provided the full config-x64-windows-rel-err.log for you.

Here it is again:
```

Running configuration tests...
Project ERROR: Unknown feature object c++14 in expression 'features.c++14 || tests.cxx14_make_unique'.
```

**Comment by JackBoosY** [comment]
Those 2 options are qt config options, see https://interest.qt-project.narkive.com/xv113Rd6/qmake-and-config-c-14.
If qt cannot recognize them, it should be that qmake.exe does not match the qt config helper script provided by qt5-base.
This means that the qt5-base you have installed has expired.
In addition, both Visual Studio 2019 and Visual Studio 2022 support cxx14, so this should have nothing to do with the compiler.

**Comment by dannypike** [comment]
OK, we're going around in circles, so let's please start this again "from the top" on the assumption that I am not making a silly error and neither are you. So, with apologies for the length of this message but I want you to read what I am saying and not what you think that I am saying.

1. Since you said that you cannot reproduce the problem, I have been using completely new virtual machines installed with the latest version of Windows 10 to test it. There is no "infection" from old versions of anything, nothing can have expired because everything is the latest version: Windows, VS2019, VS2022, vcpkg and the qt5 port.
2. I am entering one command only on that brand-new, clean, uncorrupted and unedited version of an installation of `vcpkg`:
```
vcpkg install qt5-declarative
```
That command fails with the error listed above.

From what I see here, is nothing else running, installed, edited or anything that could have invalidated the test. If it works for you, then I maintain that the problem is that you have something installed on your machine that is not part of the `qt5-declarative` port but was recently added to the `qt5-imageformats` port, as that one suddenly started building OK when I updated my `vcpkg` installation on 13 Nov, when it had previously been exhibiting *exactly this same problem*.

It is my current assertion that a brand new clean unadulterated version of the current `qt5` port *will not build* on a clean unadulterated version of Visual Studio 2019 *or* 2022. It used to build just fine (I have been using Qt5 for years) and is was broken by a recent change in some part of the setup (Visual Studio, vcpkg, qt5, cmake or whatever).

I don't know or particularly care which bit of the infrastructure is causing it to fail, but I do want it to be fixed, please. My suggestion is that you look at whatever was recently done to `qt5-imageformats` and do the same to `qt5-declarative` and, perhaps, some of the other Qt5 ports. This problem may then go away and we can get on with our lives.

Thanks and sorry for the rant, but this seems to be taking weeks to get over the "but it works for me, so it must be your fault" virtual wall that you have built. I am not responsible for the Windows, VS2019, VS2022 or vcpkg source code, so I maintain that it is not my fault. If it will not compile with brand-new installations of everything, then it is not my fault that it does not compile!

**Comment by JackBoosY** [comment]
@dannypike Sorry about that, I did say that it's yout fault.

@Neumann-A Any ideas here?

**Comment by Neumann-A** [comment]
> @Neumann-A Any ideas here?

if the build is successful in CI and there is no repro I cannot come to any conclusion why that might be happening. maybe build/install with `--x-cmake-args=--trace-expand` to see which qmake is used.  `qt5-declarative` however might fail on longer paths.

**Comment by JackBoosY** [comment]
So please provide the output using the following command:
`.\vcpkg install qt5-imageformats:x64-windows --x-cmake-args=--trace-expand`.
Please note the output is huge and you should put it into a file.

**Comment by dannypike** [comment]
@JackBoosY Before I do that, can you please tell me why are you still talking about `qt5-imageformats`?

**Comment by LilyWangLL** [comment]
@dannypike I still can't repro this issue locally, I installed Visual Studio 16.11.7, and I can install qt5-declarative:x64-windows succeed. This is my environment information:
![image](https://user-images.githubusercontent.com/94091114/143156415-59af43fd-17fc-4f3c-bef6-413ea267da90.png)
![image](https://user-images.githubusercontent.com/94091114/143156450-7649e399-82bf-49fe-bd0b-edd8210693eb.png)

**Comment by JackBoosY** [comment]
Sorry, because English is not my native language, my understanding may be partly biased.
The command should be `.\vcpkg install qt5-declarative:x64-windows --x-cmake-args=--trace-expand`.

**Comment by dannypike** [comment]
@LilyWangLL Thank you very much for the details.

For me, the error happens with a Hyper-V virtual machine (VM) that is running `Community 2019 16.11.7` and another VM that is running `Community 2022 17.0.1`.

It does not happen with a VM that is running `Community 2019 16.11.5` and, from your experience, with one that is running `Enterprise 2019 Preview 16.11.7`.

It seems to be very sensitive to the installation.

**Comment by dannypike** [comment]
@JackBoosY Thank you for the clarification. I was just making sure that we are talking about the same module.

I did not install `qmake` manually. Qt5 is installed by `vcpkg` and the VM is a new one, so it has not seen anything other versions.

The output from trying to install `qt5-declarative` with tracing enabled is 2800 lines long:

[vcpkg_qt5-declarative-trace.txt](https://github.com/microsoft/vcpkg/files/7593260/vcpkg_qt5-declarative-trace.txt)

Here is the beginning of the __Command Prompt__ output, which shows the command that I used to create that file:

```
C:\work\github\vcpkg>vcpkg install qt5-declarative:x64-windows --x-cmake-args=--trace-expand | tee vcpkg_qt5-declarative-trace.txt
Computing installation plan...
The following packages will be built and installed:
    qt5-declarative[core]:x64-windows -> 5.15.2#2
Restored 0 packages from C:\Users\Nevyn\AppData\Local\vcpkg\archives in 152.5 us. Use --debug to see more details.
Starting package 1/1: qt5-declarative:x64-windows
Building package qt5-declarative[core]:x64-windows...
Running with expanded trace output on.
C:/work/github/vcpkg/scripts/ports.cmake(1):  cmake_minimum_required(VERSION 3.21 )
C:/work/github/vcpkg/scripts/ports.cmake(3):  set(SCRIPTS C:/work/github/vcpkg/scripts CACHE PATH Location to stored scripts )
C:/work/github/vcpkg/scripts/ports.cmake(4):  list(APPEND CMAKE_MODULE_PATH C:/work/github/vcpkg/scripts/cmake )
C:/work/github/vcpkg/scripts/ports.cmake(5):  include(C:/work/github/vcpkg/scripts/cmake/execute_process.cmake )
C:/work/github/vcpkg/scripts/cmake/execute_process.cmake(10):  if(NOT DEFINED Z_VCPKG_OVERRIDEN_EXECUTE_PROCESS )
.
.
.
```

and, to answer your question about which `qmake` I am using, line 2642 of the trace output file looks like this:

```
C:/work/github/vcpkg/scripts/cmake/vcpkg_configure_qmake.cmake(87):  vcpkg_execute_required_process(COMMAND C:/work/github/vcpkg/installed/x64-windows/tools/qt5/bin/qmake.exe CONFIG-=debug CONFIG+=release CONFIG-=static;CONFIG*=shared  C:/work/github/vcpkg/buildtrees/qt5-declarative/src/5.15.2-30e50540ea -qtconf C:/work/github/vcpkg/buildtrees/qt5-declarative/x64-windows-rel/qt.conf  WORKING_DIRECTORY C:/work/github/vcpkg/buildtrees/qt5-declarative/x64-windows-rel LOGNAME config-x64-windows-rel )
```

and this is the version of that `qmake` instance:
```
C:\work\github\vcpkg>C:/work/github/vcpkg/installed/x64-windows/tools/qt5/bin/qmake.exe --version
QMake version 3.1
Using Qt version 5.15.2 in C:/work/github/vcpkg/installed/x64-windows/lib
```

Finally, here is the `config-x64-windows-rel-err.log` from that attempt to install `qt5-declarative`:
[config-x64-windows-rel-err.log](https://github.com/microsoft/vcpkg/files/7593307/config-x64-windows-rel-err.log)

which is only those same three lines that I have given you before:

```
C:/work/github/vcpkg/buildtrees/qt5-declarative/src/5.15.2-30e50540ea/.qmake.conf:1: Cannot find feature qt_build_config
Could not find qmake spec 'win32-msvc'.
Error processing project file: C:\work\github\vcpkg\buildtrees\qt5-declarative\src\5.15.2-30e50540ea\qtdeclarative.pro
```

The VM is still available, so please let me know if you need any other information from it.

**Comment by JackBoosY** [comment]
Related: https://bugreports.qt.io/browse/QTVSADDINBUG-675.
I wish this issue is not related to the file permission.

**Comment by dannypike** [comment]
I've just tried another `git pull && vcpkg upgrade`, which updated lots of ports; now `qt5-imageformats` doesn't build any more:

```
C:/work/github/vcpkg/buildtrees/qt5-imageformats/src/5.15.2-bdc20f8631.clean/.qmake.conf:1: Cannot find feature qt_build_config
Could not find qmake spec 'win32-msvc'.
Error processing project file: C:\work\github\vcpkg\buildtrees\qt5-imageformats\src\5.15.2-bdc20f8631.clean\qtimageformats.pro
```

That is also now complaining about feature `qt_build_config`, rather than the `c++14` one. The error appears to change every time the Qt5 port is upgraded. If there is an important file installed on your machine that isn't in the standard `vcpkg` distribution, then [the ticket](https://bugreports.qt.io/browse/QTVSADDINBUG-675) that you mentioned might lead to the behaviour we are seeing here.

If this hypothesis is correct, `qmake` may be doing a "initial check" by looking in several places for some important files and failing the build if it cannot find one of them, although it may not actually use that same map to load those files. Does that make sense to you?

This is a possible route to the error:

1. `qmake` builds a map of file names that it thinks that it will need later on. (It must be a __map__ and not a __list__, because the name of the "missing" feature keeps changing.)
2. `qmake` searches through the map looking for these files and aborts the build when it thinks that cannot find one of them in the hard-coded folder that it is looking at. (This is what I see.)
3. Otherwise, `qmake` carries on configuring the build script. (This is what you see.)
4. `qmake` runs the build script, __which uses a different list of paths to load the files__ from the `vcpkg` folders.
5. Everything builds (for you).

Does that make sense? If `Qt5` is checking the files that it needs by using a different algorithm from the one that it uses to load those files, then that is a bug in `Qt5` and it may cause this annoying behaviour. One way that we can test this is for you to use a blank installation that has never seen any earlier installation of anything, especially anything that involves `Qt5`.

So, please would you try a test using a new Hyper-V virtual machine with new installations of `git`, `vcpkg` and the compiler?

**Comment by LilyWangLL** [comment]
`qt_build_config.prf` exists in _VCPKG_ROOT\installed\TRIPLET\tools\qt5\mkspecs\features_.
`win32-msvc` config files are in __VCPKG_ROOT\installed\TRIPLET\tools\qt5\mkspecs_.
Those 2 paths should not be set in the qmake.exe command line.

In _vcpkg\buildtrees\qt5-declarative\internal-config-x64-windows-dbg.log_:
```
> 	cl -c -nologo -Zc:wchar_t -FS -Zc:rvalueCast -Zc:inline -Zc:strictStrings -Zc:throwingNew -Zc:referenceBinding -Zc:__cplusplus -Zi -MDd -W0 -EHsc /Fdpointer_64bit.vc.pdb -DUNICODE -D_UNICODE -DWIN32 -D_ENABLE_EXTENDED_ALIGNED_STORAGE -DWIN64 -I. -IE:\1114\vcpkg\installed\x64-windows\include -IE:\1114\vcpkg\installed\x64-windows\include\qt5 -IE:\1114\vcpkg\installed\x64-windows\tools\qt5\debug\mkspecs\win32-msvc -Fo @C:\Users\USR\AppData\Local\Temp\nm8B7E.tmp
```

Can you please provide all the log files which in directory _C:\work\github\vcpkg\buildtrees\qt5-declarative_?

Thanks.

**Comment by dannypike** [comment]
@LilyWangLL Here is a zipped archive of everything in `buildtrees/qt5-declarative` except the `src` subdirectory:

[qt5-declarative.zip](https://github.com/microsoft/vcpkg/files/7593886/qt5-declarative.zip)

**Comment by dannypike** [comment]
@LilyWangLL Here are the locations of the configuration files that you mentioned:

```
Microsoft Windows [Version 10.0.19043.1348]
(c) Microsoft Corporation. All rights reserved.

C:\Users\Nevyn>cd \

C:\>dir /s /b qt_build_config.prf
C:\work\github\vcpkg\buildtrees\qt5-base\src\5.15.2-8e14085175.clean\mkspecs\features\qt_build_config.prf
C:\work\github\vcpkg\installed\x64-windows\tools\qt5\debug\mkspecs\features\qt_build_config.prf
C:\work\github\vcpkg\installed\x64-windows\tools\qt5\mkspecs\features\qt_build_config.prf
C:\work\github\vcpkg\packages\qt5-base_x64-windows\tools\qt5\debug\mkspecs\features\qt_build_config.prf
C:\work\github\vcpkg\packages\qt5-base_x64-windows\tools\qt5\mkspecs\features\qt_build_config.prf

C:\>dir /s /b win32-msvc*
C:\work\github\vcpkg\buildtrees\qt5-base\src\5.15.2-8e14085175.clean\mkspecs\win32-msvc
C:\work\github\vcpkg\installed\x64-windows\tools\qt5\debug\mkspecs\win32-msvc
C:\work\github\vcpkg\installed\x64-windows\tools\qt5\mkspecs\win32-msvc
C:\work\github\vcpkg\packages\qt5-base_x64-windows\tools\qt5\debug\mkspecs\win32-msvc
C:\work\github\vcpkg\packages\qt5-base_x64-windows\tools\qt5\mkspecs\win32-msvc

C:\>

```

If `qmake` has a bug, then it will not matter whether I have the files in the *correct* directory. The problem might be that I don't have some files in the *wrong* directory as well. :)

This is why I think you need to try this on a new installation (a VM).

**Comment by LilyWangLL** [comment]
The following content is extra in _buildtrees/qt5-declarative/x64-windows-rel/qt.conf_:
```
[EffectiveSourcePaths]
Prefix=U:/qt5-base/src/5.15.2-8e14085175.clean
```
@Neumann-A Any ideas?

**Comment by JackBoosY** [comment]
> Seems like:
> 
> https://github.com/microsoft/vcpkg/blob/a2fcb03749ff5897b5985092934dc6057680c789/ports/qt5-base/cmake/configure_qt.cmake#L145
> 
> 
> is not working.

Because the path is not in vcpkg.
That's why qt5-declarative has configure error here.

**Comment by dannypike** [comment]
I'm sorry, everyone. The confusion over the U:/ folder is my fault.

I was in a hurry yesterday morning and I tried to save time by copying some of the packages from my main machine on to the VM. I was not thinking clearly enough, and did not consider the possibility that the Qt port configuration might use fully-rooted path names.

My main machine uses batch files to control `vcpkg`, and those batch files use the `--x-buildtrees` option with some `subst` drives to reduce the path names in the builds. The blank VMs don't do that, because its not needed for qt5 (only for qt6), but it explains why the error I get on my main machine is about a missing "c++14 feature", whereas the error on the VMs is about missing "qt build config".

I will need to do these tests again, which will take some time as building Qt on virtual machines is a very slow process. Please give me a day or two to get back to you. It's looking hopeful, though. I now know a lot more about how to analyse the error report myself.

**Comment by JackBoosY** [comment]
Using --x-buildtrees will definitely destroy the configuration process of qt, @Neumann-A should we take this into consideration?

**Comment by dannypike** [comment]
Could you use relative path names to find the Qt installation?

I would be quite happy for the Qt configuration to require that all of the packages are based off the same `vcpkg\installed\x64-windows` folder, and I do use the VCPKG_ROOT to find that, myself. However, I was a little bit surprised that it uses a full pathname. It feels "wrong" to have to keep the `subst` drives active forever.

**Comment by JackBoosY** [comment]
> Could you use relative path names to find the Qt installation?
> 
> I would be quite happy for the Qt configuration to require that all of the packages are based off the same `vcpkg\installed\x64-windows` folder, and I do use the VCPKG_ROOT to find that, myself. However, I was a little bit surprised that it uses a full pathname. It feels "wrong" to have to keep the `subst` drives active forever.

I think this is a vcpkg bug and we will fix that to solve this.

**Comment by Neumann-A** [comment]
I think replacing it with https://github.com/microsoft/vcpkg/issues/21656#issuecomment-978935249 should be enough.

you can try to manually remove 
```
[EffectiveSourcePaths]
Prefix=U:/qt5-base/src/5.15.2-8e14085175.clean
```

>  and did not consider the possibility that the Qt port configuration might use fully-rooted path names

It shouldn't. the qt.conf in tools/qt5/bin should be relativ. The qt.conf in the buildtree is generated before configure with full paths so every build should have the correct paths. the `[EffectiveSourcePaths]` stuff is just something which did not correctly removed.

**Comment by dannypike** [comment]
> the qt.conf in tools/qt5/bin should be relativ.

OK, it's good to know that you feel the same way about using relative paths for everything, thank you.

Now that I know what's going wrong, I have managed to get my build with `VS2022 17.0.1` to complete. I am currently running a build with `VS2019 16.11.7`. If that also works, then I will be happy to close this ticket and I'm happy to wait for you to fix the underlying problem with the `qt` scripts, as part of  #21656.

**Comment by JackBoosY** [comment]
I will open a PR to fix this using @Neumann-A suggestion later.

**Comment by BillyONeal** [comment]
> My main machine uses batch files to control `vcpkg`, and those batch files use the `--x-buildtrees` option with some `subst` drives to reduce the path names in the builds. The blank VMs don't do that, because its not needed for qt5 (only for qt6), but it explains why the error I get on my main machine is about a missing "c++14 feature", whereas the error on the VMs is about missing "qt build config".

Can you show the command which produced these? There's a find and replace that attempts to remove the path roots from the qt files which may be tripped up if the *string* of the path is different even if the semantics of the path are the same. For example, using S:/different\\slashes/and\\CaSeS

**Comment by dannypike** [comment]
@BillyONeal I believe you are correct and I think that's what they are trying to fix with [this PR](https://github.com/microsoft/vcpkg/pull/21657) (although it's not passing the CI tests, yet).

**Comment by JackBoosY** [comment]
@Neumann-A The only place calls `configure_qt` is in qt5-base's portfile.cmake, which means that other qt ports will not do these replacements.
